### ISSUE TITLE
Make directives module or class

### ISSUE THREAD

**Issue by b-butler** [glotzerlab/signac-flow#265]
### Feature description
Directives are the core of being able to computationally difficult jobs whether the problem is approached with MPI, OPENMP, or other parallelization schemes. Currently this logic is primarily in `FlowGroup`. Moving this to a module or just helper functions that take lists of directives or a single operation's directive, and computes correct aggregation, properly applies callable directives, etc. would be useful.

Some ideas for functionality (additions and critiques welcome):
* Aggregate directives in parallel and serial
* Check common typos (eg `ngpu` v `ngpus`)
* Get a set of unused directives with whitelisting and blacklisting handled by the function/class

### Proposed solution

I am not sure whether a class or function approach would be better. I am open to either, but I think this would make the code more modular, simple, and extensible.

**Comment by kidrahahjo** [comment]
@csadorf @b-butler 
Can I try to resolve this issue?

**Comment by b-butler** [comment]
@kidrahahjo I appreciate the enthusiasm, but this task will requires familiarity with cluster submission. I think that it may be best to tackle another issue.

**Comment by kidrahahjo** [comment]
@b-butler  
Okay.
I think I'll follow the complete pull request which fixes this issue in order to gain more knowledge on cluster submission so that in the near future I can myself solve these kinds of issues.
Thank you

**Comment by b-butler** [comment]
At a recent dev meeting we discussed the outline for what this should look like. We have concluded that directives should be their own class. In general, this class should be a mapping from string keys (e.g. `nranks`, `ngpu`, `executable`) to sensible values for each particular key (to match the previous example `int`, `int`, `str`). We also need logic to combine multiple directive instances into a single instance for both parallel and serial execution. After some discussion @vyasr and I have concluded that having a structure like what follows makes sense for our needs (our conversation did not discuss all the following details but the gist is there).

````python
class DirectivesItem:
    def __init__(self, name, type, serial_agg_method, parallel_agg_method, validation_func, ...):
        pass

class Directives:
    def __init__(self, directives_item_list):
        pass

# ...

class ComputeEnvironment:
    def _get_directives_instance(self):
        return Directives([DirectiveItems('ngpu', int, max, sum, lambda x: x > 0)])
````

Upon the registering of operations, groups, labels, etc. before `FlowProject` instantiation, directives would be stored as dictionaries. Upon the instantiation of the project, we could convert the registered directives to instances of `Directives` provided by the `FlowProject`'s environment. The `Directives` class would have the logic necessary to aggregate directives serially or parallel as defined by the options given at construction. The `Directives` interface should be similar to a `dict` with more error checking for keys that are meaningful to signac internally.

One thing we did not discuss and is worthy of mention here, is that directives like `executable` cannot be aggregated in any meaningful way and generally don't fit the pattern of directives as resource requests.

**Comment by csadorf** [comment]
Maybe allow the provision of `None` as an aggregation function, which means "cannot aggregate" this specific directive in this way. That would be the default for all directives that are not handled.

**Comment by ac-optimus** [comment]
hi, can I work on resolving this issue? This is the first step to solve this [issue](https://github.com/glotzerlab/signac-flow/issues/270). Thanks!

**Comment by b-butler** [comment]
@ac-optimus are you familiar with submitting jobs to remote clusters? The initial states of this at least require some familiarity with submitting jobs to a cluster.

**Comment by ac-optimus** [comment]
I have worked with simple-scheduler while dealing with previous issues. I find this issue interesting, if required I can also set up a small one-node SLURM on my linux as once suggested by @bdice on slack channel?
